google/cloud/opentelemetry/internal/monitored_resource_test.cc

```diff
@@ -0,0 +1,321 @@
+      {"us-central1-a", absl::nullopt, "us-central1-a"},
+      {absl::nullopt, "us-central1", "us-central1"},
+  };
+  for (auto const& test : tests) {
+    auto attributes = opentelemetry::sdk::resource::ResourceAttributes{
+        {sc::kCloudPlatform, "gcp_kubernetes_engine"},
+        {sc::kK8sClusterName, "test-cluster"},
+        {sc::kK8sNamespaceName, "test-namespace"},
+        {sc::kK8sPodName, "test-pod"},
+        //{sc::kK8sContainerName, "test-container"},
+    };
+    if (test.zone) attributes[sc::kCloudAvailabilityZone] = *test.zone;
+    if (test.region) attributes[sc::kCloudRegion] = *test.region;
+
+    auto mr = ToMonitoredResource(attributes);
+    EXPECT_EQ(mr.type, "k8s_pod");
+    EXPECT_THAT(mr.labels,
+                UnorderedElementsAre(Pair("location", test.expected_location),
+                                     Pair("cluster_name", "test-cluster"),
+                                     Pair("namespace_name", "test-namespace"),
+                                     Pair("pod_name", "test-pod")));
+  }
+}
+
+TEST(MonitoredResource, K8sNode) {
+  struct TestCase {
+    absl::optional<std::string> zone;
+    absl::optional<std::string> region;
+    std::string expected_location;
+  };
+  auto tests = std::vector<TestCase>{
+      {"us-central1-a", "us-central1", "us-central1-a"},
+      {"us-central1-a", absl::nullopt, "us-central1-a"},
+      {absl::nullopt, "us-central1", "us-central1"},
+  };
+  for (auto const& test : tests) {
+    auto attributes = opentelemetry::sdk::resource::ResourceAttributes{
+        {sc::kCloudPlatform, "gcp_kubernetes_engine"},
+        {sc::kK8sClusterName, "test-cluster"},
+        {sc::kK8sNodeName, "test-node"},
+        //{sc::kK8sPodName, "test-pod"},
+        //{sc::kK8sContainerName, "test-container"},
+    };
+    if (test.zone) attributes[sc::kCloudAvailabilityZone] = *test.zone;
+    if (test.region) attributes[sc::kCloudRegion] = *test.region;
+
+    auto mr = ToMonitoredResource(attributes);
+    EXPECT_EQ(mr.type, "k8s_node");
+    EXPECT_THAT(mr.labels,
+                UnorderedElementsAre(Pair("location", test.expected_location),
+                                     Pair("cluster_name", "test-cluster"),
+                                     Pair("node_name", "test-node")));
+  }
+}
+
+TEST(MonitoredResource, K8sCluster) {
+  struct TestCase {
+    absl::optional<std::string> zone;
+    absl::optional<std::string> region;
+    std::string expected_location;
+  };
+  auto tests = std::vector<TestCase>{
+      {"us-central1-a", "us-central1", "us-central1-a"},
+      {"us-central1-a", absl::nullopt, "us-central1-a"},
+      {absl::nullopt, "us-central1", "us-central1"},
+  };
+  for (auto const& test : tests) {
+    auto attributes = opentelemetry::sdk::resource::ResourceAttributes{
+        {sc::kCloudPlatform, "gcp_kubernetes_engine"},
+        {sc::kK8sClusterName, "test-cluster"},
+        //{sc::kK8sNodeName, "test-node"},
```

Comment:
Why are these commented out?

Reply:
I was being cheeky. I thought it would better convey the conditions of these if statement: https://github.com/googleapis/google-cloud-cpp/blob/6b87fd6484997467a219be2a30d5a41519493433/google/cloud/opentelemetry/internal/monitored_resource.cc#L185-L194

I was inspired by this: https://github.com/googleapis/google-cloud-cpp/blob/0ca17bf611207ab8e2db84d4ceb539e0acdfafa9/google/cloud/internal/oauth2_external_account_credentials_test.cc#L404-L419

Anyways, I removed them.